File: mRemoteNG/Forms/PasswordForm.vb
Imports mRemoteNG.My

Namespace Forms
    Public Class PasswordForm
#Region "Public Properties"
        Public Property Verify As Boolean = True

        Public ReadOnly Property Password As String
            Get
                If Verify Then
                    Return txtVerify.Text
                Else
                    Return txtPassword.Text
                End If
            End Get
        End Property
#End Region


#Region "Constructors"
        Public Sub New(Optional ByVal passwordName As String = Nothing, Optional ByVal verify As Boolean = True)
            ' This call is required by the designer.
            InitializeComponent()

            ' Add any initialization after the InitializeComponent() call.
            _passwordName = passwordName
            Me.Verify = verify
        End Sub
#End Region

#Region "Event Handlers"
        Private Sub frmPassword_Load(ByVal sender As Object, ByVal e As EventArgs) Handles Me.Load
            ApplyLanguage()

            If Not Verify Then
                Height = Height - (txtVerify.Top - txtPassword.Top)
                lblVerify.Visible = False
                txtVerify.Visible = False
            End If
        End Sub

        Private Sub btnCancel_Click(ByVal sender As System.Object, ByVal e As EventArgs) Handles btnCancel.Click
            DialogResult = DialogResult.Cancel
        End Sub

        Private Sub btnOK_Click(ByVal sender As System.Object, ByVal e As EventArgs) Handles btnOK.Click
            If Verify Then
                If VerifyPassword() Then DialogResult = DialogResult.OK
            Else
                DialogResult = DialogResult.OK
            End If
        End Sub

        Private Sub txtPassword_TextChanged(sender As System.Object, e As EventArgs) Handles txtPassword.TextChanged, txtVerify.TextChanged
            HideStatus()
        End Sub
#End Region

#Region "Private Fields"
        Private ReadOnly _passwordName As String
#End Region

#Region "Private Methods"
        Private Sub ApplyLanguage()
            If String.IsNullOrEmpty(_passwordName) Then
                Text = Language.strTitlePassword
            Else
                Text = String.Format(Language.strTitlePasswordWithName, _passwordName)
            End If

            lblPassword.Text = Language.strLabelPassword
            lblVerify.Text = Language.strLabelVerify

            btnCancel.Text = Language.strButtonCancel
            btnOK.Text = Language.strButtonOK
        End Sub

        Private Function VerifyPassword() As Boolean
            If txtPassword.Text.Length >= 3 Then
                If txtPassword.Text = txtVerify.Text Then
                    Return True
                Else
                    ShowStatus(Language.strPasswordStatusMustMatch)
                    Return False
                End If
            Else
                ShowStatus(Language.strPasswordStatusTooShort)
                Return False
            End If
        End Function

        Private Sub ShowStatus(ByVal status As String)
            lblStatus.Visible = True
            lblStatus.Text = status
        End Sub

        Private Sub HideStatus()
            lblStatus.Visible = False
        End Sub
#End Region
    End Class
End Namespace

File: mRemoteV1/UI/Forms/PasswordForm.cs
using System;
using System.Security;
using System.Windows.Forms;
using mRemoteNG.Security;
using mRemoteNG.Tools;

namespace mRemoteNG.UI.Forms
{
    public partial class PasswordForm : IKeyProvider
	{
        private readonly string _passwordName;
        private SecureString _password = new SecureString();

        /// <summary>
        /// Puts the dialog into the New Password mode. An extra
        /// password box is shown which must match the first password
        /// to continue.
        /// </summary>
	    private bool NewPasswordMode { get; }

        /// <summary>
        /// Creates a new password form for entering or setting a password.
        /// </summary>
        /// <param name="passwordName"></param>
        /// <param name="newPasswordMode">
        /// Puts the dialog into the New Password mode. An extra
        /// password box is shown which must match the first password
        /// to continue.
        /// </param>
		public PasswordForm(string passwordName = null, bool newPasswordMode = true)
		{
			InitializeComponent();
			_passwordName = passwordName;
			NewPasswordMode = newPasswordMode;
		}

        /// <summary>
        /// Dispaly a dialog box requesting that the user 
        /// enter their password.
        /// </summary>
        /// <returns></returns>
        public Optional<SecureString> GetKey()
        {
            var dialog = ShowDialog();
            return dialog == DialogResult.OK
                ? _password
                : Optional<SecureString>.Empty;
        }

        #region Event Handlers
        private void frmPassword_Load(object sender, EventArgs e)
		{
			ApplyLanguage();

		    if (NewPasswordMode) return;
		    Height = Height - (txtVerify.Top - txtPassword.Top);
		    lblVerify.Visible = false;
		    txtVerify.Visible = false;
		}

        private void PasswordForm_FormClosed(object sender, FormClosedEventArgs e)
        {
            _password = txtPassword.Text.ConvertToSecureString();
            txtPassword.Text = "";
            txtVerify.Text = "";
            if (NewPasswordMode) return;
            Height = Height + (txtVerify.Top - txtPassword.Top);
        }

        private void btnCancel_Click(object sender, EventArgs e)
		{
			DialogResult = DialogResult.Cancel;
            Close();
		}

	    private void btnOK_Click(object sender, EventArgs e)
	    {
            if (NewPasswordMode)
	            VerifyNewPassword();

	        DialogResult = DialogResult.OK;
	    }

	    private void txtPassword_TextChanged(object sender, EventArgs e)
		{
			HideStatus();
		}
        #endregion
		
        #region Private Methods
		private void ApplyLanguage()
		{
			Text = string.IsNullOrEmpty(_passwordName) ? Language.strTitlePassword : string.Format(Language.strTitlePasswordWithName, _passwordName);
				
			lblPassword.Text = Language.strLabelPassword;
			lblVerify.Text = Language.strLabelVerify;
			btnCancel.Text = Language.strButtonCancel;
			btnOK.Text = Language.strButtonOK;
		}
			
		private bool VerifyNewPassword()
		{
			if (txtPassword.Text.Length >= 3)
			{
				if (txtPassword.Text == txtVerify.Text)
					return true;
			    ShowStatus(Language.strPasswordStatusMustMatch);
			    return false;
			}
		    ShowStatus(Language.strPasswordStatusTooShort);
		    return false;
		}
			
		private void ShowStatus(string status)
		{
			lblStatus.Visible = true;
			lblStatus.Text = status;
		}
			
		private void HideStatus()
		{
			lblStatus.Visible = false;
		}
        #endregion
	}
}